File: ios/FaceBlood/FaceBlood/PulseAmplifier.swift
//
//  PulseAmplifier.swift
//  FaceBlood
//
//  GPU-accelerated colour amplification for the three pulse modes.
//
//  - SUBTLE  : soft radial tint inside the ROI (CIRadialGradient + screen blend)
//  - VIVID   : larger radial tint + soft-light wash over the whole frame
//  - EXTREME : per-pixel skin shift via a CIColorKernel — pixels detected as
//              skin are mixed dramatically toward the target colour and have
//              their saturation boosted, then composited with "lighter" blend.
//
//  The kernel implements the same heuristic as the Web build:
//    isSkin = brightness in [55, 245] and R > G > B*0.85 and R > B and R > G*1.04
//

import CoreImage
import CoreImage.CIFilterBuiltins
import UIKit
import simd

final class PulseAmplifier {

    // Reused across calls.
    private let ciContext: CIContext
    private let extremeKernel: CIColorKernel?

    init() {
        // Prefer Metal, fall back to whatever is available.
        if let device = MTLCreateSystemDefaultDevice() {
            self.ciContext = CIContext(mtlDevice: device,
                                       options: [.cacheIntermediates: false])
        } else {
            self.ciContext = CIContext(options: [.cacheIntermediates: false])
        }
        self.extremeKernel = PulseAmplifier.compileExtremeKernel()
    }

    var context: CIContext { ciContext }

    // MARK: - Public entry point
    /// - Parameters:
    ///   - input: BGRA CIImage straight from the camera.
    ///   - mode:  Selected visualization mode.
    ///   - pulse: rPPG instantaneous waveform value (-1..1 or so).
    ///   - amp:   User AMP slider value (1..10).
    ///   - confidence: 0..1, used to fade the effect on/off smoothly.
    /// - Returns: A new CIImage with the colour amplification applied.
    func process(input: CIImage,
                 mode: PulseMode,
                 pulse: Float,
                 amp: Float,
                 confidence: Float) -> CIImage {

        // Effective intensity, mirroring the JS computation.
        let intensity = max(-1.5, min(1.5, pulse * (amp / 3) * mode.ampScale))
        let sign: Float = intensity >= 0 ? 1 : -1
        let absI = min(Float(1.2), abs(intensity))
        let target = sign > 0 ? mode.systoleRGB : mode.diastoleRGB

        let effectAlpha = max(0, min(1, (confidence - 0.1) / 0.4))
        if effectAlpha <= 0.001 {
            return input
        }

        let extent = input.extent
        let centre = CGPoint(x: extent.midX, y: extent.midY)
        let shorter = min(extent.width, extent.height)
        let roi = shorter * 0.55

        switch mode.id {
        case .subtle:
            return subtleOverlay(on: input,
                                 centre: centre,
                                 roi: roi,
                                 colour: target,
                                 absI: absI * effectAlpha)
        case .vivid:
            return vividOverlay(on: input,
                                centre: centre,
                                roi: roi,
                                colour: target,
                                absI: absI * effectAlpha)
        case .extreme:
            return extremeOverlay(on: input,
                                  colour: target,
                                  absI: absI * effectAlpha)
        }
    }

    // MARK: - SUBTLE
    private func subtleOverlay(on input: CIImage,
                               centre: CGPoint,
                               roi: CGFloat,
                               colour: SIMD3<Float>,
                               absI: Float) -> CIImage {
        let alpha = CGFloat(min(0.7, absI * 0.5))
        guard let radial = CIFilter(name: "CIRadialGradient",
                                    parameters: [
                                        "inputCenter": CIVector(cgPoint: centre),
                                        "inputRadius0": 0,
                                        "inputRadius1": roi * 0.6,
                                        "inputColor0": CIColor(
                                            red: CGFloat(colour.x),
                                            green: CGFloat(colour.y),
                                            blue: CGFloat(colour.z),
                                            alpha: alpha),
                                        "inputColor1": CIColor(
                                            red: CGFloat(colour.x),
                                            green: CGFloat(colour.y),
                                            blue: CGFloat(colour.z),
                                            alpha: 0),
                                    ])?.outputImage?.cropped(to: input.extent)
        else { return input }
        return screenBlend(top: radial, base: input)
    }

    // MARK: - VIVID
    private func vividOverlay(on input: CIImage,
                              centre: CGPoint,
                              roi: CGFloat,
                              colour: SIMD3<Float>,
                              absI: Float) -> CIImage {
        let a1 = CGFloat(min(0.95, absI * 0.85))
        guard let radial = CIFilter(name: "CIRadialGradient",
                                    parameters: [
                                        "inputCenter": CIVector(cgPoint: centre),
                                        "inputRadius0": roi * 0.1,
                                        "inputRadius1": roi * 0.95,
                                        "inputColor0": CIColor(
                                            red: CGFloat(colour.x),
                                            green: CGFloat(colour.y),
                                            blue: CGFloat(colour.z),
                                            alpha: a1),
                                        "inputColor1": CIColor(
                                            red: CGFloat(colour.x),
                                            green: CGFloat(colour.y),
                                            blue: CGFloat(colour.z),
                                            alpha: 0),
                                    ])?.outputImage?.cropped(to: input.extent)
        else { return input }
        let withRadial = screenBlend(top: radial, base: input)

        // Soft-light wash over the whole frame.
        let washAlpha = CGFloat(min(0.35, absI * 0.35))
        let wash = CIImage(color: CIColor(
            red: CGFloat(colour.x),
            green: CGFloat(colour.y),
            blue: CGFloat(colour.z),
            alpha: washAlpha
        )).cropped(to: input.extent)
        return softLightBlend(top: wash, base: withRadial)
    }

    // MARK: - EXTREME
    private func extremeOverlay(on input: CIImage,
                                colour: SIMD3<Float>,
                                absI: Float) -> CIImage {
        guard let kernel = extremeKernel else {
            return input
        }
        let mixBase = min(0.95, 0.55 + absI * 0.6)
        let satBoost = 0.4 * absI
        let glowAlpha = CGFloat(min(0.45, absI * 0.30))

        let args: [Any] = [
            input,
            CIVector(x: CGFloat(colour.x), y: CGFloat(colour.y), z: CGFloat(colour.z)),
            CGFloat(mixBase),
            CGFloat(satBoost),
        ]
        guard let recoloured = kernel.apply(extent: input.extent,
                                            arguments: args) else {
            return input
        }
        // Composite with "lighter" blend (additive clipped).
        let withSkin = additiveBlend(top: recoloured, base: input)

        // Final full-screen colour glow.
        let glow = CIImage(color: CIColor(
            red: CGFloat(colour.x),
            green: CGFloat(colour.y),
            blue: CGFloat(colour.z),
            alpha: glowAlpha
        )).cropped(to: input.extent)
        return screenBlend(top: glow, base: withSkin)
    }

    // MARK: - Blend helpers
    private func screenBlend(top: CIImage, base: CIImage) -> CIImage {
        let f = CIFilter.screenBlendMode()
        f.inputImage = top
        f.backgroundImage = base
        return f.outputImage ?? base
    }
    private func softLightBlend(top: CIImage, base: CIImage) -> CIImage {
        let f = CIFilter.softLightBlendMode()
        f.inputImage = top
        f.backgroundImage = base
        return f.outputImage ?? base
    }
    private func additiveBlend(top: CIImage, base: CIImage) -> CIImage {
        let f = CIFilter.additionCompositing()
        f.inputImage = top
        f.backgroundImage = base
        return f.outputImage ?? base
    }

    // MARK: - Custom kernel
    /// CIKernel source — same skin-shift algorithm as the JS EXTREME mode but
    /// running on the GPU per fragment.
    private static let extremeKernelSource = """
    kernel vec4 skinShift(__sample s, vec3 target, float mixBase, float satBoost) {
        float r = s.r;
        float g = s.g;
        float b = s.b;
        float bright = (r + g + b) / 3.0;
        // Same heuristic as the JS build, normalized to 0..1.
        float minB = 55.0 / 255.0;
        float maxB = 245.0 / 255.0;
        float skin = step(minB, bright)
                   * step(bright, maxB)
                   * step(g, r)
                   * step(b * 0.85, g)
                   * step(b, r)
                   * step(g * 1.04, r * 1.0);
        if (skin < 0.5) {
            return s;
        }
        // Weight: brighter skin gets stronger shift.
        float weight = clamp((bright - 50.0/255.0) / (180.0/255.0), 0.0, 1.0);
        float k = mixBase * weight;
        vec3 mixed;
        mixed.r = min(1.0, r * (1.0 - k) + target.r * k);
        mixed.g = min(1.0, g * (1.0 - k) + target.g * k);
        mixed.b = min(1.0, b * (1.0 - k) + target.b * k);
        // Saturation boost.
        float avg = (mixed.r + mixed.g + mixed.b) / 3.0;
        mixed.r = min(1.0, mixed.r + (mixed.r - avg) * satBoost);
        mixed.g = min(1.0, mixed.g + (mixed.g - avg) * satBoost);
        mixed.b = min(1.0, mixed.b + (mixed.b - avg) * satBoost);
        return vec4(mixed, s.a);
    }
    """

    private static func compileExtremeKernel() -> CIColorKernel? {
        return CIColorKernel(source: extremeKernelSource)
    }
}

File: android/app/src/main/java/com/nwlab/faceblood/service/PulseAmplifier.kt
package com.nwlab.faceblood.service

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.RadialGradient
import android.graphics.Shader
import com.nwlab.faceblood.model.PulseMode
import com.nwlab.faceblood.model.PulseModeId
import kotlin.math.abs
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Software colour amplification for the three pulse modes.
 *
 * Mirrors the iOS [PulseAmplifier] and the Web canvas overlay logic:
 *
 *  - SUBTLE  : soft radial tint inside the ROI (screen blend).
 *  - VIVID   : larger radial tint + soft-light wash over the whole frame.
 *  - EXTREME : per-pixel skin-colour shift with saturation boost.
 *
 * All operations run on a CPU [Bitmap] / [Canvas].  For production use the
 * RenderScript / AGSL path would be preferable, but this keeps the code
 * dependency-free and easy to read in Android Studio.
 */
class PulseAmplifier {

    /**
     * Apply the colour amplification effect to [input] and return a new [Bitmap].
     *
     * @param input      Camera frame (ARGB_8888).
     * @param mode       Selected visualization mode.
     * @param pulse      rPPG instantaneous waveform value (~-1..1).
     * @param amp        User AMP slider value (1..10).
     * @param confidence Confidence score 0..1 — used to fade the effect in/out.
     */
    fun process(
        input: Bitmap,
        mode: PulseMode,
        pulse: Float,
        amp: Float,
        confidence: Float,
    ): Bitmap {
        // Effective intensity — mirrors the JS / Swift computation.
        val intensity = (pulse * (amp / 3f) * mode.ampScale).coerceIn(-1.5f, 1.5f)
        val sign = if (intensity >= 0f) 1f else -1f
        val absI = min(1.2f, abs(intensity))
        val targetR: Float
        val targetG: Float
        val targetB: Float
        if (sign > 0f) {
            targetR = mode.systoleR; targetG = mode.systoleG; targetB = mode.systoleB
        } else {
            targetR = mode.diastoleR; targetG = mode.diastoleG; targetB = mode.diastoleB
        }

        val effectAlpha = ((confidence - 0.1f) / 0.4f).coerceIn(0f, 1f)
        if (effectAlpha <= 0.001f) return input

        val w = input.width
        val h = input.height
        val cx = w / 2f
        val cy = h / 2f
        val shorter = min(w, h).toFloat()
        val roi = shorter * 0.55f

        return when (mode.id) {
            PulseModeId.SUBTLE  -> subtleOverlay(input, cx, cy, roi, targetR, targetG, targetB, absI * effectAlpha)
            PulseModeId.VIVID   -> vividOverlay(input, cx, cy, roi, targetR, targetG, targetB, absI * effectAlpha)
            PulseModeId.EXTREME -> extremeOverlay(input, targetR, targetG, targetB, absI * effectAlpha)
        }
    }

    // -------------------------------------------------------------------------
    // SUBTLE
    // -------------------------------------------------------------------------

    private fun subtleOverlay(
        base: Bitmap,
        cx: Float, cy: Float, roi: Float,
        r: Float, g: Float, b: Float,
        absI: Float,
    ): Bitmap {
        val alpha = min(0.7f, absI * 0.5f)
        val out = base.copy(Bitmap.Config.ARGB_8888, true)
        val canvas = Canvas(out)
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            shader = RadialGradient(
                cx, cy, roi * 0.6f,
                intArrayOf(
                    argb(alpha, r, g, b),
                    argb(0f, r, g, b),
                ),
                null,
                Shader.TileMode.CLAMP,
            )
            xfermode = PorterDuffXfermode(PorterDuff.Mode.SCREEN)
        }
        canvas.drawRect(0f, 0f, base.width.toFloat(), base.height.toFloat(), paint)
        return out
    }

    // -------------------------------------------------------------------------
    // VIVID
    // -------------------------------------------------------------------------

    private fun vividOverlay(
        base: Bitmap,
        cx: Float, cy: Float, roi: Float,
        r: Float, g: Float, b: Float,
        absI: Float,
    ): Bitmap {
        val a1 = min(0.95f, absI * 0.85f)
        val out = base.copy(Bitmap.Config.ARGB_8888, true)
        val canvas = Canvas(out)

        // Radial gradient (screen blend)
        val radialPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            shader = RadialGradient(
                cx, cy, roi * 0.95f,
                intArrayOf(
                    argb(a1, r, g, b),
                    argb(0f, r, g, b),
                ),
                floatArrayOf(roi * 0.1f / (roi * 0.95f), 1f),
                Shader.TileMode.CLAMP,
            )
            xfermode = PorterDuffXfermode(PorterDuff.Mode.SCREEN)
        }
        canvas.drawRect(0f, 0f, base.width.toFloat(), base.height.toFloat(), radialPaint)

        // Soft-light wash over the whole frame
        val washAlpha = min(0.35f, absI * 0.35f)
        val washPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = argb(washAlpha, r, g, b)
            xfermode = PorterDuffXfermode(PorterDuff.Mode.SOFT_LIGHT)
        }
        canvas.drawRect(0f, 0f, base.width.toFloat(), base.height.toFloat(), washPaint)
        return out
    }

    // -------------------------------------------------------------------------
    // EXTREME — per-pixel skin shift
    // -------------------------------------------------------------------------

    private fun extremeOverlay(
        base: Bitmap,
        r: Float, g: Float, b: Float,
        absI: Float,
    ): Bitmap {
        val mixBase = min(0.95f, 0.55f + absI * 0.6f)
        val satBoost = 0.4f * absI
        val glowAlpha = min(0.45f, absI * 0.30f)

        // Per-pixel skin shift
        val pixels = IntArray(base.width * base.height)
        base.getPixels(pixels, 0, base.width, 0, 0, base.width, base.height)
        skinShiftPixels(pixels, r, g, b, mixBase, satBoost)

        val out = Bitmap.createBitmap(base.width, base.height, Bitmap.Config.ARGB_8888)
        out.setPixels(pixels, 0, base.width, 0, 0, base.width, base.height)

        // Additive composite (lighter blend)
        val canvas = Canvas(out)
        val addPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            xfermode = PorterDuffXfermode(PorterDuff.Mode.ADD)
        }
        canvas.drawBitmap(base, 0f, 0f, addPaint)

        // Full-screen colour glow (screen blend)
        val glowPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = argb(glowAlpha, r, g, b)
            xfermode = PorterDuffXfermode(PorterDuff.Mode.SCREEN)
        }
        canvas.drawRect(0f, 0f, base.width.toFloat(), base.height.toFloat(), glowPaint)
        return out
    }

    /**
     * Applies the same skin-detection heuristic as the iOS CIColorKernel and
     * the Web JS EXTREME mode, but running on the CPU per pixel.
     *
     * Skin heuristic (normalised 0..1):
     *   brightness in [55/255, 245/255]  AND  R > G  AND  G > B*0.85  AND  R > B  AND  R > G*1.04
     */
    private fun skinShiftPixels(
        pixels: IntArray,
        tR: Float, tG: Float, tB: Float,
        mixBase: Float, satBoost: Float,
    ) {
        val minB = 55f / 255f
        val maxB = 245f / 255f
        for (i in pixels.indices) {
            val px = pixels[i]
            val pr = ((px shr 16) and 0xFF) / 255f
            val pg = ((px shr 8) and 0xFF) / 255f
            val pb = (px and 0xFF) / 255f
            val pa = (px ushr 24) and 0xFF

            val bright = (pr + pg + pb) / 3f
            val isSkin = bright in minB..maxB &&
                pr > pg &&
                pg > pb * 0.85f &&
                pr > pb &&
                pr > pg * 1.04f

            if (!isSkin) continue

            val weight = ((bright - 50f / 255f) / (180f / 255f)).coerceIn(0f, 1f)
            val k = mixBase * weight
            var mr = min(1f, pr * (1f - k) + tR * k)
            var mg = min(1f, pg * (1f - k) + tG * k)
            var mb = min(1f, pb * (1f - k) + tB * k)

            // Saturation boost
            val avg = (mr + mg + mb) / 3f
            mr = min(1f, mr + (mr - avg) * satBoost)
            mg = min(1f, mg + (mg - avg) * satBoost)
            mb = min(1f, mb + (mb - avg) * satBoost)

            pixels[i] = (pa shl 24) or
                ((mr * 255f).toInt().coerceIn(0, 255) shl 16) or
                ((mg * 255f).toInt().coerceIn(0, 255) shl 8) or
                (mb * 255f).toInt().coerceIn(0, 255)
        }
    }

    // -------------------------------------------------------------------------
    // Colour helpers
    // -------------------------------------------------------------------------

    /** Pack float RGBA (0..1) into an Android ARGB int. */
    private fun argb(a: Float, r: Float, g: Float, b: Float): Int =
        ((a * 255f).toInt().coerceIn(0, 255) shl 24) or
            ((r * 255f).toInt().coerceIn(0, 255) shl 16) or
            ((g * 255f).toInt().coerceIn(0, 255) shl 8) or
            (b * 255f).toInt().coerceIn(0, 255)
}
